worker/upload.go

```diff
@@ -0,0 +1,1154 @@
+
+	// create a host
+	h, err := hp.newHostV3(j.requestCtx, j.queue.fcid, j.queue.hk, j.queue.siamuxAddr)
+	if err != nil {
+		return types.Hash256{}, err
+	}
+
+	// upload the sector
+	start := time.Now()
+	root, err := h.UploadSector(j.requestCtx, j.sector, rev)
+	if err != nil {
+		return types.Hash256{}, err
+	}
+
+	// update span
+	elapsed := time.Since(start)
+	span.SetAttributes(attribute.Int64("duration", elapsed.Milliseconds()))
+	span.RecordError(err)
+	span.End()
+	return root, nil
+}
+
+func (j *uploadJob) succeed(root types.Hash256) {
+	select {
+	case <-j.requestCtx.Done():
+	case j.responseChan <- sectorResponse{
+		job:  j,
+		root: root,
+	}:
+	}
+}
+
+func (j *uploadJob) fail(err error) {
+	select {
+	case <-j.requestCtx.Done():
+	case j.responseChan <- sectorResponse{
+		job: j,
+		err: err,
+	}:
+	}
+}
+
+func (j *uploadJob) done() bool {
+	select {
+	case <-j.requestCtx.Done():
+		return true
+	default:
+		return false
+	}
+}
+
+func (s *uploadState) launch(job *uploadJob) error {
+	s.mu.Lock()
+	defer s.mu.Unlock()
+
+	// enqueue the job
+	err := s.u.enqueue(job)
+	if err != nil {
+		span := trace.SpanFromContext(job.requestCtx)
+		span.RecordError(err)
+		span.End()
+		return err
+	}
+
+	// update the state
+	s.numInflight++
+	s.numLaunched++
+	if job.overdrive {
+		s.lastOverdrive = time.Now()
+		s.overdriving[job.sectorIndex]++
+	}
+
+	return nil
+}
+
+func (s *uploadState) receive(resp sectorResponse) (completed bool) {
+	s.mu.Lock()
+	defer s.mu.Unlock()
+
+	// failed jobs can't complete the upload
+	s.numInflight--
+	if resp.err != nil {
+		s.errs = append(s.errs, &HostError{resp.job.queue.hk, resp.err})
+		return false
+	}
+
+	defer func() {
+		s.u.registerCompletedSector(s.uploadID, s.shardID, resp.job.queue.fcid, completed)
+	}()
+
+	// redundant sectors can't complete the upload
+	s.numCompleted++
+	if s.sectors[resp.job.sectorIndex].Root != (types.Hash256{}) {
+		return false
+	}
+
+	// store the sector and call cancel on the sector ctx
+	s.sectors[resp.job.sectorIndex] = object.Sector{
+		Host: resp.job.queue.hk,
+		Root: resp.root,
+	}
+	s.remaining[resp.job.sectorIndex].cancel()
+
+	// count the sector as complete and check if we're done
+	delete(s.remaining, resp.job.sectorIndex)
+
+	// trigger next slab read if we only have some outstanding sectors left
+	if !s.nextReadTriggered && len(s.remaining) < int(s.u.maxOverdrive) {
+		s.nextReadTriggered = true
+		s.u.triggerNextSlab(s.uploadID)
+	}
+
+	if len(s.remaining)%5 == 0 || len(s.remaining) < 5 {
+		fmt.Printf("DEBUG PJ: %v | %v | remaining sectors %d\n", s.uploadID, s.shardID, len(s.remaining))
+	}
+	return len(s.remaining) == 0
+}
+
+func (s *uploadState) inflight() uint64 {
+	s.mu.Lock()
+	defer s.mu.Unlock()
+	return s.numInflight
+}
+
+func (s *uploadState) performance() float64 {
+	s.mu.Lock()
+	defer s.mu.Unlock()
+	downloaded := int64(s.numCompleted) * rhpv2.SectorSize
+	elapsed := time.Since(s.created)
+	return mbps(downloaded, elapsed.Seconds())
```

Comment:
I would probably drop as many `float64` as possible and use `bytes / elapsed.Milliseconds()` as `uint64` here.
Floats are incredibly unreliable and I'd avoid them for as long as possible until we absolutely need to return mbps.